fontique/src/backend/dwrite.rs

```diff
@@ -120,129 +113,270 @@ impl SystemFonts {
+            let factory2: IDWriteFactory2 = factory.cast().ok()?;
+            let fallback = factory2.GetSystemFontFallback().ok()?;
+            Some(Self {
+                collection,
+                fallback,
+                map_buf: vec![],
+            })
+        }
+    }
+
+    fn get(&self, index: u32) -> Option<DWriteFontFamily> {
+        unsafe {
+            self.collection
+                .GetFontFamily(index)
+                .ok()
+                .map(DWriteFontFamily)
+        }
+    }
+
+    fn family_by_name(&self, name: &str) -> Option<DWriteFontFamily> {
+        let mut index = 0;
+        let mut exists = Default::default();
+        let mut name_buf: smallvec::SmallVec<[u16; 128]> = Default::default();
+        name_buf.extend(name.encode_utf16());
+        name_buf.push(0);
+        unsafe {
+            self.collection
+                .FindFamilyName(PCWSTR::from_raw(name_buf.as_ptr()), &mut index, &mut exists)
+                .ok()?;
+        }
+        self.get(index as _)
+    }
+
+    fn families(&self) -> impl Iterator<Item = DWriteFontFamily> {
+        let this = self.collection.clone();
+        unsafe {
+            let count = this.GetFontFamilyCount();
+            (0..count).filter_map(move |index| this.GetFontFamily(index).ok().map(DWriteFontFamily))
+        }
+    }
+
+    fn family_name_for_text(&mut self, text: &str, locale: Option<&str>) -> Option<String> {
+        self.map_buf.clear();
+        self.map_buf.extend(text.encode_utf16());
+        let text_len = self.map_buf.len();
+        if let Some(locale) = locale {
+            self.map_buf.extend(locale.encode_utf16());
+        }
+        self.map_buf.push(0);
+        let (text, locale) = self.map_buf.split_at(text_len);
+        let source: IDWriteTextAnalysisSource = TextSource { text, locale }.into();
+        unsafe {
+            let mut cur_offset = 0;
+            while cur_offset < text_len {
+                let mut mapped_len = 0;
+                let mut mapped_font = None;
+                if self
+                    .fallback
+                    .MapCharacters(
+                        &source,
+                        cur_offset as u32,
+                        (text_len - cur_offset) as u32,
+                        &self.collection,
+                        None,
                         DWRITE_FONT_WEIGHT_REGULAR,
                         DWRITE_FONT_STYLE_NORMAL,
                         DWRITE_FONT_STRETCH_NORMAL,
-                        &mut mapped_length,
-                        &mut font,
-                        &mut scale,
+                        &mut mapped_len,
+                        &mut mapped_font,
+                        &mut 1.0,
                     )
-                };
-                assert_eq!(hr, 0);
-                if font.is_null() {
-                    i += 1;
+                    .is_ok()
+                {
+                    if let Some(font) = mapped_font {
```

Comment:
What happens if there isn't a mapped font?  I assume falling through to `None` case is fine, but the `mapped_len` doesn't seem to be used, instead just skipping a single character.

Reply:
In reality, we’re not making full use of this function. What we actually want is the font DirectWrite would choose for a specific Unicode script and locale. Since there’s no API for this, we cheat by using a table containing sample text for each script (see https://github.com/linebender/parley/blob/caabcf155dc3b8fa43ff4e78945bb87b4c5d8eac/fontique/src/script.rs#L90). That table was generated naively from the UCD and ends up with some characters that DW won’t map so this code walks the given string until a valid font is returned.

Ideally, we’d hand curate the sample table for comprehensive coverage, but so far it’s seemed to work on all platforms (with some tiny hacks like this one) so I haven’t bothered.

Reply:
That's a great explanation and a subset of this would be a great thing to have in a comment in this code. :)

Reply:
Fair point! I’ll add a comment describing this glorious hack to hopefully avoid confusion for future readers.

Reply:
@xorgy It would be good if some comment about this made its way in.